packages/flutter_tts/tizen/src/text_to_speach.cc

```diff
@@ -0,0 +1,158 @@
+// Copyright 2021 Samsung Electronics Co., Ltd. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "text_to_speach.h"
+
+#include "log.h"
+
+namespace {
+
+void OnStateChanged(tts_h tts, tts_state_e previous, tts_state_e current,
+                    void *user_data) {
+  TextToSpeach *self = static_cast<TextToSpeach *>(user_data);
+  self->OnStateChanged(previous, current);
+}
+
+void OnUtteranceCompleted(tts_h tts, int utt_id, void *user_data) {
+  TextToSpeach *self = static_cast<TextToSpeach *>(user_data);
+  self->OnUtteranceCompleted(utt_id);
+  self->Stop();
+}
+
+void OnError(tts_h tts, int utt_id, tts_error_e reason, void *user_data) {
+  TextToSpeach *self = static_cast<TextToSpeach *>(user_data);
+  self->OnError(utt_id, reason);
+}
+
+}  // namespace
+
+void TextToSpeach::Init() {
+  int ret = tts_create(&tts_);
+  if (ret != TTS_ERROR_NONE) {
+    tts_ = nullptr;
+    LOG_ERROR("[TTS] tts_create failed: %s", get_error_message(ret));
+    // TODO : throw error
+  }
+}
+
+void TextToSpeach::Deinit() {
+  if (tts_) {
+    tts_destroy(tts_);
+    tts_ = nullptr;
+  }
+}
+
+void TextToSpeach::Prepare() {
+  char *language = nullptr;
+  int voice_type = 0;
+  int ret = tts_get_default_voice(tts_, &language, &voice_type);
+  if (ret == TTS_ERROR_NONE) {
+    default_language_ = language;
+    default_voice_type_ = voice_type;
+    free(language);
+  }
+
+  ret = tts_prepare(tts_);
+  if (ret != TTS_ERROR_NONE) {
+    LOG_ERROR("[TTS] tts_prepare failed: %s", get_error_message(ret));
+    // TODO : throw error
+  }
+}
+
+void TextToSpeach::RegisterTtsCallback() {
+  // TODO : check return and throw error
+  tts_set_state_changed_cb(tts_, ::OnStateChanged, (void *)this);
+  tts_set_utterance_completed_cb(tts_, ::OnUtteranceCompleted, (void *)this);
+  tts_set_error_cb(tts_, ::OnError, (void *)this);
+}
+
+void TextToSpeach::UnregisterTtsCallback() {
+  tts_unset_state_changed_cb(tts_);
+  tts_unset_utterance_completed_cb(tts_);
+  tts_unset_error_cb(tts_);
+}
+
+void TextToSpeach::OnStateChanged(tts_state_e previous, tts_state_e current) {
+  on_state_changed_(previous, current);
+}
+
+std::vector<std::string> &TextToSpeach::GetSupportedLanaguages() {
+  if (supported_lanaguages_.size() == 0) {
+    tts_foreach_supported_voices(
```

Comment:
Is this a synchronous operation?

Reply:
I couldn't find any information regarding this, but I read the implementation briefly... 
it seems to be sync call and the actual behavior is also observed as a sync call. 
I am disappointed with the documentation of Tizen :(

https://docs.tizen.org/application/native/api/mobile/4.0/group__CAPI__UIX__TTS__MODULE.html#gae25cd6960ef5241137e19e2a70835fd5

Reply:
Relying on an undocumented behavior is not a good practice but I'm okay with it.